src/lib/coda_lib/coda_lib.ml

```diff
@@ -506,25 +506,58 @@ module Make (Inputs : Inputs_intf) = struct
     ; time_controller: Time.Controller.t
     ; snark_work_fee: Currency.Fee.t }
 
+  let peek_frontier frontier_ref =
+    Mvar.peek frontier_ref
+    |> Result.of_option
+         ~error:
+           (Error.of_string
+              "Cannot retrieve transition frontier now. Bootstrapping right \
+               now.")
+
   let run_snark_worker t = t.run_snark_worker
 
   let propose_keypair t = t.propose_keypair
 
-  let best_tip t = Transition_frontier.best_tip t.transition_frontier
+  let best_tip_opt t =
+    let open Option.Let_syntax in
+    let%map frontier = Mvar.peek t.transition_frontier in
+    Transition_frontier.best_tip frontier
 
-  let best_staged_ledger t =
-    Transition_frontier.Breadcrumb.staged_ledger (best_tip t)
+  let best_staged_ledger_opt t =
+    let open Option.Let_syntax in
+    let%map tip = best_tip_opt t in
+    Transition_frontier.Breadcrumb.staged_ledger tip
 
-  let best_protocol_state t =
-    Transition_frontier.Breadcrumb.transition_with_hash (best_tip t)
+  let best_protocol_state_opt t =
+    let open Option.Let_syntax in
+    let%map tip = best_tip_opt t in
+    Transition_frontier.Breadcrumb.transition_with_hash tip
     |> With_hash.data |> External_transition.Verified.protocol_state
 
-  let best_ledger t = Staged_ledger.ledger (best_staged_ledger t)
+  let best_ledger_opt t =
+    let open Option.Let_syntax in
+    let%map staged_ledger = best_staged_ledger_opt t in
+    Staged_ledger.ledger staged_ledger
+
+  let of_option value =
+    Option.value_map value ~default:`Bootstrapping ~f:(fun x ->
+        `Participating x )
+
+  let best_tip = Fn.compose of_option best_tip_opt
+
+  let best_staged_ledger = Fn.compose of_option best_staged_ledger_opt
+
+  let best_protocol_state = Fn.compose of_option best_protocol_state_opt
+
+  let best_ledger = Fn.compose of_option best_ledger_opt
```

Comment:
the expression `Fn.compose of_option` is repeated 4 times, could factor into a function